src/notebookrenderer/widgetManager.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  IComm,
+  IKernelConnection
+} from '@jupyterlab/services/lib/kernel/kernel';
+import { ICommOpenMsg } from '@jupyterlab/services/lib/kernel/messages';
+import { JupyterCadModel } from '../model';
+import {
+  IJupyterCadWidgetModelRegistry,
+  IJupyterCadWidgetManager
+} from './token';
+import { WebSocketProvider } from '@jupyterlab/docprovider';
+import { YCommProvider } from './yCommProvider';
+import { IJupyterCadModel } from '../types';
+
+const Y_DOCUMENT_PROVIDER_URL = 'api/yjs';
+
+export class JupyterCadWidgetManager implements IJupyterCadWidgetManager {
+  constructor(options: { manager: ServiceManager.IManager }) {
+    this._manager = options.manager;
+  }
+  registerKernel(kernel: IKernelConnection): void {
+    const wm = new WidgetModelRegistry({ kernel, manager: this._manager });
+    this._registry.set(kernel.id, wm);
+  }
+
+  unregisterKernel(kernelId?: string | null): void {
+    if (kernelId) {
+      this._registry.delete(kernelId);
+    }
+  }
+
+  getWidgetModel(
+    kernelId: string,
+    commId: string
+  ): IJupyterCadModel | undefined {
+    return this._registry.get(kernelId)?.getModel(commId);
+  }
+
+  private _registry = new Map<string, IJupyterCadWidgetModelRegistry>();
+  private _manager: ServiceManager.IManager;
+}
+
+export class WidgetModelRegistry implements IJupyterCadWidgetModelRegistry {
+  constructor(options: {
+    kernel: IKernelConnection;
+    manager: ServiceManager.IManager;
+  }) {
+    const { kernel, manager } = options;
+    this._manager = manager;
+    kernel.registerCommTarget('@jupytercad:widget', this._handle_comm_open);
+  }
+
+  getModel(id: string): IJupyterCadModel | undefined {
+    return this._jcadModels.get(id);
+  }
+
+  /**
+   * Handle when a comm is opened.
+   */
+  private _handle_comm_open = async (
+    comm: IComm,
+    msg: ICommOpenMsg
+  ): Promise<void> => {
+    const { path, format, contentType } = msg.content.data as {
+      path?: string;
+      format?: string;
+      contentType?: string;
+    };
+
+    const jcadModel = new JupyterCadModel({});
+    const user = this._manager.user;
+    if (path && format && contentType) {
+      const server = ServerConnection.makeSettings();
+      const serverUrl = URLExt.join(server.wsUrl, Y_DOCUMENT_PROVIDER_URL);
+      const ywsProvider = new WebSocketProvider({
+        url: serverUrl,
+        path,
+        format,
+        contentType,
+        model: jcadModel.sharedModel,
+        user
+      });
+      jcadModel.disposed.connect(() => {
+        ywsProvider.dispose();
+      });
+      await ywsProvider.ready;
+    } else {
+      const awareness = jcadModel.sharedModel.awareness;
+      const _onUserChanged = (user: User.IManager) => {
+        awareness.setLocalStateField('user', user.identity);
+      };
+      user.ready
+        .then(() => {
+          _onUserChanged(user);
+        })
+        .catch(e => console.error(e));
+      user.userChanged.connect(_onUserChanged, this);
```

Comment:
Why is this code here? I don't really understand what it has to do with the rest of the `_handle_comm_open` method.

Reply:
Basically, this snippet is done by `WebSocketProvider`, in the case of opening a new widget without providing the file path, we don't initialize the `WebSocketProvider` but `jcadModels` (and then the 3D viewer) still expects that its awareness has the user information.